File: ans.cpp
#include<bits/stdc++.h>
using namespace std;

map<string,int>getWordFrequencies(const string &filePath){
    ifstream file(filePath);
    map<string,int>frequencies;
    string word;
    if (file.is_open()) {
        while (file >> word) {
            frequencies[word]++;
        }
        file.close();
    } else {
        cerr << "Error 404!!! Can't open file" << endl;
    }
    return frequencies;
}

vector<pair<string,int>>selectionSort(const map<string,int> &frequencies){
    vector<pair<string,int>>wordFreq(frequencies.begin(),frequencies.end());
    int n = wordFreq.size();
    for(int i=0;i<n-1;i++){
        int maxIdx = i;
        for(int j = i+1;j<n;j++){
            if (wordFreq[j].second>wordFreq[maxIdx].second){
                maxIdx = j;
            }
        }
        swap(wordFreq[i],wordFreq[maxIdx]);
    }
    return wordFreq;
}
//SELECTION SORT
pair<string,int>usingSelectionSort(const map<string,int> &frequencies,int k){
    vector<pair<string,int>>sortedWords = selectionSort(frequencies);
    return sortedWords[k - 1]; 
}

int partition(vector<pair<string,int>> &arr,int low,int high){
    int pivot = arr[high].second; 
    int i = low - 1;
    for(int j=low;j<high;j++){
        if(arr[j].second >= pivot){ 
            i++;
            swap(arr[i],arr[j]);
        }
    }
    swap(arr[i+1],arr[high]);
    return i+1;
}
//QUICK SELECT
pair<string,int>quickSelect(vector<pair<string,int>> &arr,int low,int high,int k) {
    if(low<=high){
        int pivotIndex = partition(arr,low,high);
        if(pivotIndex == k){
            return arr[pivotIndex];
        }else if(pivotIndex>k){
            return quickSelect(arr,low,pivotIndex-1,k);
        }else{
            return quickSelect(arr,pivotIndex+1,high,k);
        }
    }
    return {"", -1};
}

pair<string,int>usingQuickSelectQuickSelect(const map<string,int> &frequencies,int k){
    vector<pair<string,int>>wordFreq(frequencies.begin(),frequencies.end());
    return quickSelect(wordFreq,0,wordFreq.size()-1,k-1); 
}

//HASH TABLE AND MIN Heap
pair<string,int>usingHeap(const map<string,int> &frequencies,int k){
    priority_queue<pair<int,string>,vector<pair<int,string>>,greater<pair<int,string>>>pq;
    for(const auto &entry : frequencies){
        pq.push({entry.second,entry.first});
        if(pq.size()>k){
            pq.pop();
        }
    }
    return {pq.top().second,pq.top().first}; 
}

int main(){
  string filePath = "input.txt";//intended file with input texts
  int k = 3; 
  map<string,int>frequencies = getWordFrequencies(filePath);//first we calculate the frequencies
  pair<string,int>kthWordA = usingSelectionSort(frequencies,k);
  cout << "First Answer with Selection Sort: " << kthWordA.first << " -> " << kthWordA.second << endl;
  pair<string,int>kthWordB = usingQuickSelect(frequencies, k);
  cout << "Second Answer with Quick Select: " << kthWordB.first << " -> " << kthWordB.second << endl;
  pair<string,int>kthWordC = usingHeap(frequencies, k);
  cout << "Third Answer Using Heap: " << kthWordC.first  << " -> " << kthWordC.second << endl;
  return 0;
}

File: ans.py
from collections import Counter
import heapq

# Get word frequencies from file
def get_word_frequencies(file_path):
    with open(file_path,'r') as file:
        text = file.read()
    words = text.split()  
    frequencies = Counter(words)  
    return frequencies

# Selection Sort
def selection_sort(words_with_freq):
    n = len(words_with_freq)
    for i in range(n-1):
        max_idx = i
        for j in range(i+1,n):
            if words_with_freq[j][1] > words_with_freq[max_idx][1]: 
                max_idx = j
        words_with_freq[i], words_with_freq[max_idx] = words_with_freq[max_idx], words_with_freq[i]
    return words_with_freq

def using_selection_sort(frequencies,k):
    words_with_freq = list(frequencies.items())  
    sorted_words = selection_sort(words_with_freq) 
    return sorted_words[k - 1]  

#Quick Select
def partition(arr,low,high):
    pivot = arr[high][1]  
    i = low-1
    for j in range(low,high):
        if arr[j][1]>=pivot:  
            i += 1
            arr[i],arr[j] = arr[j],arr[i]
    arr[i+1],arr[high] = arr[high],arr[i+1]
    return i + 1

def quick_select(arr,low,high,k):
    if low <= high:
        pivot_index = partition(arr,low,high)
        if pivot_index == k:
            return arr[pivot_index]
        elif pivot_index>k:
            return quick_select(arr,low,pivot_index-1,k)
        else:
            return quick_select(arr,pivot_index+1,high,k)
    return None

def using_quick_select(frequencies,k):
    words_with_freq = list(frequencies.items())  
    return quick_select(words_with_freq,0,len(words_with_freq)-1,k-1)  
#Using heap
def using_heap(frequencies,k):
    pq = []
    for word,freq in frequencies.items():
        heapq.heappush(pq,(freq,word)) 
        if len(pq)>k:
            heapq.heappop(pq)  
    top = heapq.heappop(pq)
    return (top[1],top[0]) 

# Main Function
if __name__ == "__main__":
    file_path = "input.txt"  
    k = 3  
    frequencies = get_word_frequencies(file_path)
    kth_word_a = using_selection_sort(frequencies,k)
    print(f"Using Selection Sort: {kth_word_a}")
    kth_word_b = using_quick_select(frequencies,k)
    print(f"Using Quick Select: {kth_word_b}")
    kth_word_c = using_heap(frequencies,k)
    print(f"Using Hash Table and Heap: {kth_word_c}")
